crates/matrix-sdk-crypto/src/store/mod.rs

```diff
@@ -203,16 +198,15 @@ impl<'a> SyncedKeyQueryManager<'a> {
         timeout_duration: Duration,
         user: &UserId,
     ) -> UserKeyQueryResult {
-        let mut users_for_key_query = self.manager.users_for_key_query.lock().await;
-
-        let Some(waiter) = users_for_key_query.maybe_register_waiting_task(user) else {
+        let Some(waiter) =
+            self.manager.users_for_key_query.lock().await.maybe_register_waiting_task(user)
+        else {
             return UserKeyQueryResult::WasNotPending;
         };
 
         let wait_for_completion = async {
             while !waiter.completed.load(Ordering::Relaxed) {
-                users_for_key_query =
-                    self.manager.users_for_key_query_condvar.wait(users_for_key_query).await;
+                self.manager.users_for_key_query_notify.notified().await;
```

Comment:
the problem here is that we are not holding a lock, so an update could arrive between the call to `waiter.completed.load` and the the call to `users_for_key_query_notify.notified`. In that case, we would miss the notification, and end up waiting indefinitely.

This is the point of a condvar: it atomically drops a lock while you are waiting for the event, which means you can protect surrounding code with a lock.

Reply:
Aha, I thought the point was in being able to register to be "next in line" to acquire the mutex after a certain condition. Luckily, tokio's `Notify` allows registering for wakeup before waiting, so we can replicate the same semantics. Please take a look at the [change](https://github.com/matrix-org/matrix-rust-sdk/compare/d32dbb416183136b9317b852e25df05ab8a87437..7ff5f45a7b190e41f58c9623ba9622f6f6b78771) I just pushed.